pkg/controlplane/controller/defaultservice/default_service_controller.go

```diff
@@ -0,0 +1,337 @@
+	// it will not work for clients because of the IP family mismatch.
+	// TODO: revisit for dual-stack https://github.com/kubernetes/enhancements/issues/2438
+	if endpointReconcilerType != reconcilers.NoneEndpointReconcilerType {
+		if netutils.IsIPv4(serviceIP) != netutils.IsIPv4(endpointIP) {
+			return nil, fmt.Errorf("service IP family %q must match endpoint (apiserver) address family %q", serviceIP.String(), endpointIP.String())
+		}
+	}
+
+	broadcaster := record.NewBroadcaster()
+	recorder := broadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: controllerName})
+
+	c := &Controller{
+		endpointReconciler: endpointsReconciler,
+		serviceIP:          serviceIP,
+		servicePort:        servicePort,
+		serviceNodePort:    serviceNodePort,
+		endpointIP:         endpointIP,
+		endpointPort:       endpointPort,
+		client:             client,
+		endpointInterval:   endpointInterval,
+	}
+	// we construct our own informer because we need such a small subset of the information available.
+	// The kubernetes.default service
+	serviceInformer := corev1informers.NewFilteredServiceInformer(client, metav1.NamespaceDefault, 12*time.Hour,
+		cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+		func(options *metav1.ListOptions) {
+			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", serviceName).String()
+		})
+
+	c.serviceInformer = serviceInformer
+	c.serviceLister = corev1listers.NewServiceLister(serviceInformer.GetIndexer())
+	c.servicesSynced = serviceInformer.HasSynced
+
+	c.eventBroadcaster = broadcaster
+	c.eventRecorder = recorder
+
+	return c, nil
+}
+
+// Controller manages selector-based service ipAddress.
+type Controller struct {
+	client           clientset.Interface
+	eventBroadcaster record.EventBroadcaster
+	eventRecorder    record.EventRecorder
+
+	// service frontend
+	serviceIP       net.IP
+	servicePort     int
+	serviceNodePort int
+
+	// published apiserver IP and Port
+	endpointIP   net.IP
+	endpointPort int
+
+	serviceInformer cache.SharedIndexInformer
+	serviceLister   corev1listers.ServiceLister
+	servicesSynced  cache.InformerSynced
+
+	endpointReconciler reconcilers.EndpointReconciler
+	endpointInterval   time.Duration
+
+	stopOnce sync.Once
+}
+
+// Run will not return until ctx.Done() is closed.
+func (c *Controller) start(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	c.eventBroadcaster.StartStructuredLogging(0)
+	klog.Infof("Sending events to API server.")
+	c.eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: c.client.CoreV1().Events("")})
+	defer c.eventBroadcaster.Shutdown()
+
+	klog.Infof("Starting %s", controllerName)
+	defer klog.Infof("Shutting down %s", controllerName)
+
+	// Reconcile during first run removing itself until server is ready.
+	endpointPorts := createEndpointPortSpec(c.endpointPort, "https")
+	if err := c.endpointReconciler.RemoveEndpoints(serviceName, c.endpointIP, endpointPorts); err == nil {
```

Comment:
note to myself: this is a newly added thing, but that seems desired

Reply:
it is not new https://github.com/kubernetes/kubernetes/blob/75889ecec5d30cbe1dcb6636d5334b21b4378e9c/pkg/controlplane/controller.go#L150-L156